documentation/writing_tests.md

```diff
@@ -14,6 +17,96 @@
 
 ## Recommended tools
 
-## Structure of tests
+## Test folder structure
+
+Generally speaking, tests for the Azure SDK for JavaScript and TypeScript should be all executed using the NPM command `test`, they should be stored in the `test` folder, they will be written in TypeScript, in files that will end in `.spec.ts`, and our [recommended tools](#recommended-tools) should help us with the test framework, with proper ways of compiling them and bundling them correctly, so they can run and be debugged while checking that our code works on the environments we're targeting (NodeJS, browsers, etc).
+
+To help developers specify the target environment they will be testing, we will specify these scripts in the `package.json`:
+
+```json
+  "test:browser": "karma start --single-run",
+  "test:node": "mocha -r esm --require source-map-support/register --reporter mocha-multi --reporter-options spec=-,mocha-junit-reporter=- --timeout 180000 --full-trace dist-esm/test/*.test.js",
+```
+
+Besides the target environment, tests may target live resources (or not) and public or internal code. To navigate through these concepts, we will make clear distinctions of what we will be testing, based on three main aspects:
+
+- [Testing cloud resources](#testing-cloud-resources).
+- [Public or internal tests](#public-or-internal-tests).
+- [Testing API functionalities](#testing-api-functionalities).
+
+We will continue by elaborating on these three distinctions.
+
+### Testing cloud resources
+
+To specify how the tests interact with the cloud resources, we will be using [the recorder](#the-recorder), which allows us to write the same tests once, targeting real live resources, and then run them again instantly through specifying the `TEST_MODE` environment variable to either `record`, `live` or `playback`. Not specifying the environment variable will be considered `playback`, since we want to be able to run the tests without having to set up real resources.
+
+### Public or internal tests
+
+Tests can also either target internal code or code that is part of the public API of our projects. We will separate where these tests are stored by dividing them in folders inside of the `test` folder. As follows:
+
+```
+project/
+  test/
+    internal/
+	  internal_method.spec.ts
+	public/
+	  public_method.spec.ts
+```
+
+Keep in mind that...
+
+- **Internal** tests should:
+  - Never target real resources, or resources outside of the code that exists inside of each project.
+  - Not invoke or use [the recorder](#the-recorder), nor benefit from using the recorder.
+  - Possibly use custom mocks, only purposed for these internal tests.
+  - Never target public facing API methods or properties.
+- **Public** tests should:
+  - Only target public facing API methods or properties.
+  - Use [the recorder](#the-recorder).
+  - Not use custom mocks.
+  - Use light dynamic code, and preferably be user friendly, like samples.
```

Comment:
What's an example of light dynamic code? Or do you mean something like "don't crazy with DRY?"

Reply:
That's what I mean. We shouldn't go crazy with DRY. I'll try to improve this way to express it, but suggestions are appreciated!